### ISSUE TITLE
"Update and start" is not triggering two builds. 

### ISSUE THREAD

**Issue by BrunoQuaresma** [coder/coder#13539]
When updating and starting a workspace the following builds should be triggered:
- transition=stop, template_version=old
- transition=start, template_version=new

But it looks like it is not happening on 2.12.1.

Reported by @spikecurtis

**Comment by BrunoQuaresma** [comment]
I look at the files where this could happen but looks like we don't stop the workspace before starting a new "start build" for updates and it has been like that for a while.

Places where it could be changed:
- https://github.com/coder/coder/blob/main/site/src/pages/WorkspacePage/WorkspaceReadyPage.tsx#L124-L126
- https://github.com/coder/coder/blob/main/site/src/api/queries/workspaces.ts#L141-L153
- https://github.com/coder/coder/blob/main/site/src/api/api.ts#L1697-L1734

**Comment by stirby** [comment]
Did not mean to close, misclick.

**Comment by matifali** [comment]
We can follow the same process for workspace which are pending auto deletion but have an update available.
Currently
- transition=delete, template_version=old

which fails the deletion as the old template has some values not valid anymore e.g. (the docker host address has been updated)

It should be 
- transition=start, template_version=new

**Comment by Emyrk** [comment]
@BrunoQuaresma @aslilac Do we have prior art for something like this?

Issuing a "stop" would require watching the logs, waiting for it to complete, then issuing the next api call.

This feels more involved then simplying posting 2 builds to the api.

**Comment by BrunoQuaresma** [comment]
We discussed this before, and I would prefer to resolve it by shifting the update action to the backend as a transition, rather than relying on the client side to manage it all.

**Comment by Emyrk** [comment]
@BrunoQuaresma who was in that discussion? Is there some notes anywhere? I agree it would be ideal to handle this in the backend. This could be done in a number of places though.

**Comment by BrunoQuaresma** [comment]
Not really 😢, but we've been facing recurring bugs related to workspace actions. One of the findings was that the frontend is handling too much, trying to calculate all the possibilities considering many states and configurations. It would probably be better to move this functionality to the backend.

**Comment by deansheather** [comment]
I think the issue of "updates should stop then start running workspaces" should be closed. Users should write their Terraform files in the Terraform-intended way where Terraform can detect drift and correct it (perhaps by restarting/recreating instances). We provide values that change on each build (agent ID, agent token, etc.) that users can use for change detection

The unrelated issue of "the backend should handle stop+start builds" should be moved to a separate issue.

**Comment by deansheather** [comment]
Reclassified as a feature request as this is intended behavior and not a regression or bug. I still think it should be closed for the aforementioned reason, and the docs should be improved to mention that templates must handle start -> start transitions.

**Comment by Emyrk** [comment]
> The unrelated issue of "the backend should handle stop+start builds" should be moved to a separate issue.

Agree on this. The issue of workspace state management should be scoped in another issue. 

-----

> Reclassified as a feature request as this was intended behavior and not a regression or bug. I still think it should be closed for the aforementioned reason, and the docs should be improved to mention that templates must handle start -> start transitions.

I am going to close this issue. If it is a feature request, we should add some more substance and reasoning imo. As it stands, this ticket does not have enough direction or rational to make such a change.

What we need for some issues:
- A bug report with steps to reproduce. If there is a terraform template that requires this 2 step approach (stop, then start), then that terraform should be included in the bug report and we can discuss solutions.
- A feature request that provides some reasoning/rational for the 2 step.

If either is expanded upon, then we can open a new issue.

**Comment by Kira-Pilot** [comment]
@sreya we looked into this and realized we have some inconsistency in transition behaviors for update and restart.

**update**: has always just started a new build, i.e. what is described in this ticket is not new behavior.
**restart**: the FE stops the current build and then starts a new build. This was implemented [here](https://github.com/coder/coder/pull/7301). 

We *did* have some confusing language in both the UI and docs that may have led users to believe update behavior was different and @Emyrk has removed that language [here](https://github.com/coder/coder/pull/13852/files).

We've closed this out as it doesn't seem to be a bug, but perhaps we should consider opening a new ticket to do the following:
1) understand what terraform templates actually need a stop before start because this seems to be an edge case
2) have consistency for update and restart if we think that's the right move
3) update the docs to reflect that consistency
4) move that transition logic to the BE because having it in the FE is buggy
5) consider an RFC to simplify workspace actions because we get a lot of issues related to this area
